ovs-dpdk.md

```diff
@@ -0,0 +1,227 @@
+
+Bind port to a suitable module:
+
+```
+# dpdk-devbind -b <module_name> <pci_addr>
+```
+
+### Hugepages Support
+Hugepages support is required for increasing performance. By allocating hugepages, we gain larger pages, using less pages and therefore less TLB misses are involved which speeds up performance. For these reasons, 1GB hugepages will be favored.
+Number of pages can vary according to usage requirements. 
+Hugepages must be marked as shared in order to enable host and guest communication, as packets are transmitted and received on buffers allocated on hugepages memory. 
+ 
+Add engine custom properties for hugepages support (followed by engine restart)
+```
+# engine-config -s "UserDefinedVMProperties=hugepages=[0-9];hugepages_shared=(true|false);hugepagesz=1024"
+```
+ 
+In oVirt GUI, configure hugepages size, number of pages to use and enable shared hugepages.
+
+### Enable DPDK in OVS:
+
+
+ 
+Add the following entry to  /etc/sysconfig/openvswitch:
+```
+"DPDK_OPTIONS='-c coremask 100010 -n 2 --socket-mem 1024,0”
+```
+The pmd-cpu-mask and -c coremask have to be both specified, with the same value.
+ 
+If allocated more than one GB hugepage, set amount and use NUMA node 0 memory.
+
+
+To better scale the workloads across cores, Multiple pmd threads can be created and pinned to CPU cores by explicitly specifying pmd-cpu-mask. eg: To spawn 2 pmd threads and pin them to cores 1, 2
+
+run the following commands:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem="1024,0" 
+# ovs-vsctl set Open_vSwitch . other_config:pmd-cpu-mask=6
+
+```
+
+
+Set dpdk-init so OVS would initialize and support DPDK ports:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+ 
+Restart openvswitch service
+```
+# sudo systemctl restart openvswitch
+```
+ 
+Verify DPDK is enabled:
+```
+# ovs-vsctl get Open_vSwitch . iface_types
+[dpdk, dpdkr, dpdkvhostuser, dpdkvhostuserclient, geneve, gre, internal, ipsec_gre, lisp, patch, stt, system, tap, vxlan]
+```
+
+### Make the changes persistent
+
+Create a new service unit in /etc/systemd/system/ovs-dpdk.service with content based on the following:
+
+```
+[Unit]
+Description=ovs-dpdk configuration script
+Before=openvswitch.service
+
+[Service]
+Type=simple
+ExecStart=/usr/local/sbin/ovs-dpdk.sh
+TimeoutStartSec=0
+
+[Install]
+WantedBy=default.target
+
+```
+
+
+``` /usr/local/sbin/ovs-dpdk.sh
+#!/bin/bash
+
+modprobe vfio-pci
+dpdk-devbind -u <pci_addr>
+dpdk-devbind -b vfio-pci <pci_addr>
+
+ovs-vsctl set Open_vSwitch . other_config:pmd-cpu-mask=6
+ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem="1024,0"
+ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+
+Enable the new service:
+
+```
+# systemctl enable ovs-dpdk.service
+```
+
+
+## OVS-DPDK in oVirt
+ 
+* Upon creating a new cluster, choose ovs as the switch type.
+* When configuring host network interfaces, DPDK-bound NICs will be presented in the system.
```

Comment:
above bullet is not clear.

do you mean "when setting up host networks, DPDK-bound NICs are shown in the dialog,as well as Linux-controlled ones"